operators/tokenizer/clip_tokenizer.cc

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+// Partial code comes from other Microsoft employee.
+
+#include <string>
+#include <vector>
+#include <fstream>
+#include <sstream>
+#include <iostream>
+#include <algorithm>
+#include <list>
+#include <memory>
+#include <regex>
+#include <sstream>
+#include <stdexcept>
+#include <unordered_map>
+#include <functional>
+#include <codecvt>
+#include <mutex>
+
+#include "nlohmann/json.hpp"
+#include "bpetokenizer.hpp"
+#include "string_tensor.h"
+#include "unicode.h"
+
+// Note: the following logic comes from CPython: unicodetype_db.h (_PyUnicode_IsWhitespace)
+bool IsInUnicodeSpace(char32_t ch) {
+  switch (ch) {
+    case 0x0009:
+    case 0x000A:
+    case 0x000B:
+    case 0x000C:
+    case 0x000D:
+    case 0x001C:
+    case 0x001D:
+    case 0x001E:
+    case 0x001F:
+    case 0x0020:
+    case 0x0085:
+    case 0x00A0:
+    case 0x1680:
+    case 0x2000:
+    case 0x2001:
+    case 0x2002:
+    case 0x2003:
+    case 0x2004:
+    case 0x2005:
+    case 0x2006:
+    case 0x2007:
+    case 0x2008:
+    case 0x2009:
+    case 0x200A:
+    case 0x2028:
+    case 0x2029:
+    case 0x202F:
+    case 0x205F:
+    case 0x3000:
+      return true;
+  }
+  return false;
+}
+
+bool IsEmptyUstring(const ustring& str) {
+  return std::all_of(str.begin(), str.end(), [](char32_t ch) { return IsInUnicodeSpace(ch); });
+}
+
+KernelClipBpeTokenizer::KernelClipBpeTokenizer(const OrtApi& api, const OrtKernelInfo* info)
+    : BaseKernel(api, info) {
+  std::string vocab = ort_.KernelInfoGetAttribute<std::string>(info, "vocab");
+  if (vocab.empty()) {
+    ORTX_CXX_API_THROW("vocabulary shouldn't be empty.", ORT_INVALID_ARGUMENT);
+  }
+
+  std::string merges = ort_.KernelInfoGetAttribute<std::string>(info, "merges");
+  if (merges.empty()) {
+    ORTX_CXX_API_THROW("merges shouldn't be empty.", ORT_INVALID_ARGUMENT);
+  }
+
+  if (!TryToGetAttribute<int64_t>("padding_length", padding_length_)) {
+    padding_length_ = -1;
+  }
+
+  if (padding_length_ != -1 && padding_length_ <= 0) {
+    ORTX_CXX_API_THROW("padding_length should be more than 0 or equal -1", ORT_INVALID_ARGUMENT);
+  }
+
+  std::stringstream vocabu_stream(vocab);
+  std::stringstream merges_stream(merges);
+  bbpe_tokenizer_ = std::make_shared<VocabData>();
+  bbpe_tokenizer_->Load(vocabu_stream, merges_stream, "<|endoftext|>", "<|endoftext|>");
+}
+
+std::vector<int64_t> KernelClipBpeTokenizer::Tokenize(ustring& input, int64_t max_length) {
+  std::vector<int64_t> res;
+
+  if (IsEmptyUstring(input)) {
+    return res;
+  }
+  // Add <|startoftext|> token to result
+  res.push_back(49406);
+
+  // Convert to lowercase
+  std::transform(input.begin(), input.end(), input.begin(), [](char c) { return static_cast<char>(ToLower(c)); });
```

Comment:
the character type of ustring should be char32_t, instead of 'char'. The 'static_cast' will introduce some Unicode issues here.